stp/stp_netlink.c

```diff
@@ -0,0 +1,409 @@
+/*
+ * Copyright 2019 Broadcom. The term "Broadcom" refers to Broadcom Inc. and/or
+ * its subsidiaries.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "stp_netlink.h"
+
+int stp_intf_get_netlink_fd();
+
+
+int stp_set_sock_buf_size(int sock, int optname, int size)
+{
+    int new_size = 0;
+    int old_size = 0;
+    int size_len = sizeof(size);
+    int ret = 0;
+
+    ret = getsockopt(sock, SOL_SOCKET, optname, &old_size, &size_len); 
+    if (-1 == ret)
+    {
+        STP_LOG_ERR("sock[%d] Getsockopt Failed : %s", sock, strerror(errno));
+        return -1;
+    }
+
+    ret = setsockopt(sock, SOL_SOCKET, optname, &size, size_len); 
+    if (-1 == ret)
+    {
+        STP_LOG_ERR("sock[%d] Setsockopt size : %d Failed : %s",sock, size, strerror(errno));
+        return -1;
+    }
+
+    ret = getsockopt(sock, SOL_SOCKET, optname, &new_size, &size_len); 
+    if (-1 == ret)
+    {
+        STP_LOG_ERR("sock[%d] Getsockopt Failed : %s",sock, strerror(errno));
+        return -1;
+    }
+    STP_LOG_INFO("socket[%d] buf_size old,req,new : [%d][%d][%d]", sock, old_size, size, new_size);
+    return new_size;
+}
+
+int stp_netlink_init(stp_netlink_cb_ptr *fn)
```

Comment:
On failure cases , the netlink socket needs to be closed

Reply:
Process gets terminated upon failure and sockets will be closed